File: src/main/java/org/yankov/mso/application/ui/controls/LabeledTextField.java
package org.yankov.mso.application.ui.controls;

import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import org.yankov.mso.application.UserInterfaceControls;

import java.util.function.Consumer;

public class LabeledTextField implements UserInterfaceControls {

    private String labelText;
    private Label label;
    private TextField textField;
    private VBox container;
    private String value;
    private Consumer<String> newValueConsumer;

    public LabeledTextField(String labelText, String value, Consumer<String> newValueConsumer) {
        this.labelText = labelText;
        this.container = new VBox();
        this.label = new Label();
        this.textField = new TextField();
        this.value = value;
        this.newValueConsumer = newValueConsumer;
    }

    public TextField getTextField() {
        return textField;
    }

    @Override
    public void layout() {
        label.setText(labelText);
        container.getChildren().add(label);

        textField.setText(value);
        if (newValueConsumer != null) {
            textField.textProperty().addListener((observable, oldValue, newValue) -> newValueConsumer.accept(newValue));
        }
        container.getChildren().add(textField);
    }

    @Override
    public Pane getContainer() {
        return container;
    }

}

File: src/main/scala/org/yankov/mso/application/ui/controls/LabeledTextField.scala
package org.yankov.mso.application.ui.controls

import scalafx.Includes._
import scalafx.beans.property.StringProperty
import scalafx.scene.control.{Label, TextField}
import scalafx.scene.input.KeyEvent
import scalafx.scene.layout.{Pane, VBox}

case class LabeledTextField(labelText: String,
                            value: String,
                            validator: Option[String => Boolean] = Option.empty,
                            readOnly: Boolean = false) {
  private val textField: TextField = new TextField {
    editable = !readOnly
    text = value
  }

  private val container = {
    val label = new Label {
      text = labelText
    }

    if (validator.isDefined) {
      textField.textProperty().addListener((observable, oldValue, newValue) => {
        if (validator.get.apply(newValue)) observable.asInstanceOf[StringProperty].setValue(newValue)
        else observable.asInstanceOf[StringProperty].setValue(oldValue)
      })
    }

    new VBox {
      children = List(
        label,
        textField
      )
    }
  }

  def getContainer: Pane = container

  def getValue: String = textField.getText

  def setValue(value: String): Unit = textField.setText(value)

  def setOnKeyReleased(f: KeyEvent => Unit): Unit = textField.setOnKeyReleased(f)
}
